tests/e2e/e2e_setup_test.go

```diff
@@ -326,3 +332,100 @@ func noRestart(config *docker.HostConfig) {
 		Name: "no",
 	}
 }
+
+func (s *IntegrationTestSuite) initUpgrade() {
+	for x := range s.chains {
+		c := s.chains[x]
+		s.T().Logf("submitting upgrade proposal for chain-id: %s", c.ChainMeta.Id)
+		cmdStr := fmt.Sprintf("docker exec %v bash -c 'chmod +x /scripts/osmosis_upgrade.sh && /scripts/osmosis_upgrade.sh \"%s\"'", s.valResources[c.ChainMeta.Id][0].Container.ID, c.ChainMeta.Id)
+		exec.Command("/bin/sh", "-c", cmdStr).Output()
+	}
+
+	for x := range s.chains {
+		c := s.chains[x]
+		s.T().Logf("waiting to reach upgrade height for chain-id: %s", c.ChainMeta.Id)
+		type status struct {
+			LatestHeight string `json:"latest_block_height"`
+		}
+
+		type syncInfo struct {
+			SyncInfo status `json:"SyncInfo"`
+		}
+
+		s.Require().Eventually(
+			func() bool {
+				cmdStr := fmt.Sprintf("docker exec %v bash -c 'osmosisd status'", s.valResources[c.ChainMeta.Id][0].Container.ID)
+				out, err := exec.Command("/bin/sh", "-c", cmdStr).CombinedOutput()
+				s.Require().NoError(err)
+				var syncInfo syncInfo
+				json.Unmarshal(out, &syncInfo)
+				if syncInfo.SyncInfo.LatestHeight != "75" {
+					fmt.Printf("current block height is %v, waiting for block 75\n", syncInfo.SyncInfo.LatestHeight)
+				}
+				return syncInfo.SyncInfo.LatestHeight == "75"
+			},
+			2*time.Minute,
+			5*time.Second,
+		)
+	}
+
+	for x := range s.chains {
+		c := s.chains[x]
+		for i := range c.Validators {
+			s.Require().NoError(s.dkrPool.RemoveContainerByName(s.valResources[c.ChainMeta.Id][i].Container.Name))
+		}
```

Comment:
I see a lot of for loops in this function, please add a comment before each to explain what is it that is happening there

Reply:
or a log message